### DISCUSSION TITLE
Can a resource account create and use another resource account?

### DISCUSSION THREAD

**Discussion by kshitijc1506** [aptos-labs/aptos-developer-discussions#98] [Questions] [answered]
Asked by `jojothomas` on Discord (#dev-discussion)

> Can a resource account have a resources account?
> Let say i publish with create_resource_account_and_publish, and inside my init module we i use account::create_resource_account(&signer,  SEED).
> Does it detect that it's a resource account that was used to publish it and just return the signer? Or does it create another resource account with the publishing resource account as the owner/creator/signer.

Keywords: resource account

**Comment by kshitijc1506** [comment, marked as answer]
Yes, a resource account can be used to create another resource account. 

If you publish a module under a resource account (`create_resource_account_and_publish(&signer, SEED)`), and in that module's `init_module` you call `create_resource_account(&signer, SEED)`, there is no logic that verifies that the `origin` signer reference is another resource account. You can see this [here](https://github.com/aptos-labs/aptos-core/blob/51122d9e30b0335bc2eddf51d744e810793b8417/aptos-move/framework/aptos-framework/sources/resource_account.move#L86).

Essentially, `create_resource_account(&signer, SEED)` will create a new resource account, and the original resource account will be the owner. The `signer_cap` will be stored in the original resource account's `Container` resource. You can see that logic [here](https://github.com/aptos-labs/aptos-core/blob/51122d9e30b0335bc2eddf51d744e810793b8417/aptos-move/framework/aptos-framework/sources/resource_account.move#L151).

**Reply by jojothomas1515** [reply]
Thanks for the answer.

### Another question:

when publishing with  `create_resource_account_and_publish` from aptos cli  
and i want to get the `signer_cap` for resource account and save move it to the resource account,
i used the function bellow, and i am confused as to the meaning of source address.
```rs
resource_account::retrieve_resource_account_cap(signer, @source_addr)
```
is it:
- the address of the resource account publishing the module which i assume is the signer address?
- or the address of my account in used on the aptos cli?

**Reply by kshitijc1506** [reply]
It is the address of the account that owns the `Container` resource. This would be the account used to create the resource account (address of your account used on CLI). 

Over [here](https://github.com/aptos-labs/aptos-core/blob/51122d9e30b0335bc2eddf51d744e810793b8417/aptos-move/framework/aptos-framework/sources/resource_account.move#L166), note the use of `borrow_global_mut<Container>(source_addr)`. This container is stored in the owner's account when a resource account is [created](https://github.com/aptos-labs/aptos-core/blob/51122d9e30b0335bc2eddf51d744e810793b8417/aptos-move/framework/aptos-framework/sources/resource_account.move#L147). 

The first parameter is a `&signer` reference to the resource account.
